File: algorithms/1502.can-make-arithmetic-progression-from-sequence/solution.rb
# @param {Integer[]} arr
# @return {Boolean}
def can_make_arithmetic_progression(arr)
  ## one liner ##
  # arr.sort.each_cons(2).map {|a, b| a - b }.uniq.size == 1

  ## calc each combinations ##
  arr.sort!
  diff = arr[0] - arr[1]
  (1...arr.size - 1).each do |i|
    return false unless diff == arr[i] - arr[i + 1]
  end
  # default
  true
end

File: algorithms/1502.can-make-arithmetic-progression-from-sequence/solution.rs
impl Solution {
    pub fn can_make_arithmetic_progression(arr: Vec<i32>) -> bool {
        let mut reorder = arr.clone();
        reorder.sort();

        // using HashSet //
        // use std::collections::HashSet;
        // (0..(arr.len() - 1)).map(|i| reorder[i] - reorder[i + 1]).collect::<HashSet<i32>>().len() == 1

        // calc each conbinations //
        let diff = reorder[0] - reorder[1];
        for i in 1..(arr.len() - 1) {
            if diff != reorder[i] - reorder[i + 1] {
                return false;
            }
        }

        true
    }
}
